### PR TITLE
[#4204] Fix paging results and closing statements (4-2-stable)

### PR DESCRIPTION
**Pull request by alanking**
Fixes row_idx_ going out of bounds between page_fetch() calls by
resetting to 0 in the query iterator.

Fixes stacktrace caused by closing all statements in the statement
table in the destructor of the query implementation object. This
has been changed to only close statements generated by that query.

---
[CI tests passed](http://172.25.14.63:8080/view/Personal/job/irods-build-and-test-workflow/1679/) (and [topo-resc](http://172.25.14.63:8080/job/topo-ub16-psql-resc/249/))

### CODE REVIEW

**Comment by trel** [comment]
seems good - please add #

File: lib/core/include/irods_query.hpp [hunk 1]
```diff
@@ -106,7 +106,7 @@ namespace irods {
                 return query_string_;
             }
 
-            bool page_in_flight(const int row_idx_) {
+            bool page_in_flight(const int& row_idx_) {
```
Comment (korydraughn) [inline lib/core/include/irods_query.hpp]: I don't think the `&` is really needed.  Creating a copy seems better, but the `const` pretty much achieves the same result.

File: lib/core/include/irods_query.hpp [hunk 2]
```diff
@@ -106,7 +106,7 @@ namespace irods {
...
                 if(row_idx_ < row_cnt()) {
```
Comment (korydraughn) [inline lib/core/include/irods_query.hpp]: You can replace this `if` statement with a single `return`.

File: lib/core/include/irods_query.hpp (hunk 1 above)
Comment (alanking) [reply lib/core/include/irods_query.hpp]: Got it. Thanks

File: lib/core/include/irods_query.hpp (hunk 2 above)
Comment (alanking) [reply lib/core/include/irods_query.hpp]: Done

**Comment by alanking** [comment]
[CI tests passed](http://172.25.14.63:8080/job/rule-engine-plugin-storage-tiering-build-test-workflow/237/) for storage tiering (currently the only consumer of this file)

**Pull request merged**